File: FBDeviceControl/Commands/FBDevicePowerCommands.m
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBDevicePowerCommands.h"

#import "FBDevice.h"
#import "FBAMDServiceConnection.h"

@interface FBDevicePowerCommands ()

@property (nonatomic, weak, readonly) FBDevice *device;

@end

@implementation FBDevicePowerCommands

#pragma mark Initializers

+ (instancetype)commandsWithTarget:(FBDevice *)target
{
  return [[self alloc] initWithDevice:target];
}

- (instancetype)initWithDevice:(FBDevice *)device
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _device = device;

  return self;
}

#pragma mark FBPowerCommands Implementation

- (FBFuture<NSNull *> *)shutdown
{
  return [self sendRelayCommand:@"Shutdown"];
}

- (FBFuture<NSNull *> *)reboot
{
  return [self sendRelayCommand:@"Restart"];
}

#pragma mark Private

- (FBFuture<NSNull *> *)sendRelayCommand:(NSString *)request
{
  return [[self.device
    startService:@"com.apple.mobile.diagnostics_relay"]
    onQueue:self.device.workQueue pop:^ FBFuture<NSNull *> * (FBAMDServiceConnection *connection) {
      NSError *error = nil;
      NSDictionary<NSString *, id> *result = [connection sendAndReceiveMessage:@{@"Request": request} error:&error];
      if (!result) {
        return [FBFuture futureWithError:error];
      }
      if (![result[@"Status"] isEqualToString:@"Success"]) {
        return [[FBControlCoreError
          describeFormat:@"Not successful %@", result]
          failFuture];
      }
      return FBFuture.empty;
    }];
}

@end

File: FBDeviceControl/Commands/FBDevicePowerCommands.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import FBControlCore
import Foundation

public class FBDevicePowerCommands: NSObject {
  private weak var device: FBDevice?

  // MARK: - Initializers

  public class func commands(with device: FBDevice) -> FBDevicePowerCommands {
    FBDevicePowerCommands(device: device)
  }

  init(device: FBDevice) {
    self.device = device
    super.init()
  }

  // MARK: - Private

  fileprivate func sendRelayCommand(_ request: String) async throws {
    guard let device else {
      throw FBDeviceNilError.deviceNil
    }
    try await withFBFutureContext(device.startService("com.apple.mobile.diagnostics_relay")) { connection in
      guard let result = try connection.sendAndReceiveMessage(["Request": request]) as? NSDictionary else {
        throw FBDiagnosticsRelayError.unexpectedResponse
      }
      if (result["Status"] as? String) != "Success" {
        throw FBDiagnosticsRelayError.unsuccessful(response: String(describing: result))
      }
    }
  }
}

// MARK: - FBDevice+PowerCommands

extension FBDevice: PowerCommands {

  public func shutdown() async throws {
    try await powerCommands().sendRelayCommand("Shutdown")
  }

  public func reboot() async throws {
    try await powerCommands().sendRelayCommand("Restart")
  }
}
